File: framec_tests/golang/basic/basic_test.go
package basic

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func (m *basicStruct) entered(msg string) {
	m.entry_log = append(m.entry_log, msg)
}

func (m *basicStruct) left(msg string) {
	m.exit_log = append(m.exit_log, msg)
}

func TestIntialEnterEvent(t *testing.T) {
	sm := NewBasic()
	x := sm.(*basicStruct)
	expected := []string{"S0"}
	assert.Equal(t, x.entry_log, expected)

}

func TestTransitionEnterEvents(t *testing.T) {
	sm := NewBasic()
	x := sm.(*basicStruct)
	x.entry_log = x.entry_log[:0]
	sm.A()
	sm.B()

	expected := []string{"S1", "S0"}
	assert.Equal(t, x.entry_log, expected)

}

func TestTransitionExitEvents(t *testing.T) {
	sm := NewBasic()
	x := sm.(*basicStruct)
	sm.A()
	sm.B()

	expected := []string{"S0", "S1"}
	assert.Equal(t, expected, x.exit_log)

}

func TestCurrentState(t *testing.T) {
	sm := NewBasic()
	x := sm.(*basicStruct)
	assert.Equal(t, BasicState_S0, x._compartment_.State)
	sm.A()
	assert.Equal(t, BasicState_S1, x._compartment_.State)
	sm.B()
	assert.Equal(t, BasicState_S0, x._compartment_.State)
}

File: framec_tests/cpp/basic/basic_test.cpp
#include "../gtest/gtest.h"
#include <unordered_map>
#include <stdexcept>
#include <string>
#include <iostream>
#include <vector>
using namespace std;
#include "basic.cpp"

class BasicController;  // Forward declaration

class BasicControllerTest : public ::testing::Test {
protected:
    void SetUp() override;
    void TearDown() override;

    BasicController* sm;
};

class BasicController : public Basic {
public:
    BasicController() : Basic() {}
};

void BasicControllerTest::SetUp() {
    sm = new BasicController();
}

void BasicControllerTest::TearDown() {
    delete sm;
}


TEST_F(BasicControllerTest, TestInitialEnterEvent) {
    std::vector<std::string> expected = {"S0"};
    EXPECT_EQ(expected, sm->entry_log);
}

TEST_F(BasicControllerTest, TestTransitionEnterEvents) {
    BasicController* sm = new BasicController();
    sm->entry_log.clear();
    sm->A();
    sm->B();
    std::vector<std::string> expected = {"S1", "S0"};
    ASSERT_EQ(sm->entry_log, expected);
    delete sm;
}

TEST_F(BasicControllerTest, TestTransitionExitEvents) {
    sm->A();
    sm->B();
    std::vector<std::string> expected = {"S0", "S1"};
    EXPECT_EQ(expected, sm->exit_log);
}

TEST_F(BasicControllerTest, TestCurrentState) {
    EXPECT_EQ("0", sm->state_info());
    sm->A();
    EXPECT_EQ("1", sm->state_info());
    sm->B();
    EXPECT_EQ("0", sm->state_info());
}

int main(int argc, char* argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
